src/fe/fe_base.C

```diff
@@ -2337,8 +2354,32 @@ compute_periodic_constraints (DofConstraints & constraints,
                           if (std::abs(their_dof_value) < 10*TOLERANCE)
                             continue;
 
-                          constraint_row->insert(std::make_pair(their_dof_g,
-                                                                their_dof_value));
+                          if(!periodic->has_transformation_matrix())
+                            {
+                              constraint_row->insert(std::make_pair(their_dof_g,
+                                                                    their_dof_value));
+                            }
+                          else
+                            {
+                              // In this case the current variable is constrained
+                              // in terms of other variables. We assume that all
+                              // variables in this constraint have the same FE
+                              // type, and hence we can create the constraint
+                              // row contribution by multiplying their_dof_value
```

Comment:
If we're going to assume that (which seems reasonable!), then we need to document it (in the header doxygen for set_transformation_matrix() should do it) and we need to assert it here.

Reply:
Agreed. This is what I had in mind when I said this PR needs some more work. I'll make that change.